### PR TITLE
Faster joins: don't stall when a user joins during a fast join

### LINKED ISSUE #12801: Faster joins: Handle a second join while syncing state for the first one
We want to avoid launching a second partial state syncing job for the same room.

### PR DESCRIPTION
**Pull request by MatMaul**
Fixes #12801.
Complement tests are at
https://github.com/matrix-org/complement/pull/567.

Avoid blocking on full state when handling a subsequent join into a
partial state room.

Also always perform a remote join into partial state rooms, since we do
not know whether the joining user has been banned and want to avoid
leaking history to banned users.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by MatMaul** [comment]
We had a discussion about `do_invite_join` [here](https://matrix.to/#/!vcyiEtMVHIhWXcJAfl:sw1v.org/$rN2eFSTfP5CasVf08fkhW1xV5pdpuZBM8H7Ynmx1mOc?via=matrix.org&via=element.io&via=envs.net).
TLDR:
1) this method currently removes current room extremities as a first step. If the second `send_join` is not accepted by the resident server, we don't get any new extremity and we can't send messages anymore.
2) un-partial-stating during this is not handled and should fail

For 1 we could store the membership event received from `send_join` (if succeeded) in the same transaction as deleting the old extremities.

**Comment by squahtx** [comment]
One of the ways an un-partial-stating during join could manifest is in a `PartialStateConflictError` (note the todo).
https://github.com/matrix-org/synapse/blob/54c012c5a8722725cf104fa6205f253b5b9b0192/synapse/handlers/federation.py#L643-L665

We also need to be careful that we don't return a room to partial state, because that breaks a lot of assumptions.

**Comment by MatMaul** [comment]
> 1) this method currently removes current room extremities as a first step. If the second send_join is not accepted by the resident server, we don't get any new extremity and we can't send messages anymore.
> For 1 we could store the membership event received from `send_join` (if succeeded) in the same transaction as deleting the old extremities.

It's not possible since storing events involves potentially doing so in another worker.

What I've done instead to mitigate the issue:
- do not remove forward extremities if the room is partially joined (in SQL to avoid races)
- move `_clean_room_for_join` call after `send_join`, so we avoid cleaning up if the join is failing anyway

**Comment by MatMaul** [comment]
I've added a way to fallback to local join if a room is unpartial stated during remote join.

However I am not sure this is really useful: retrying the remote join is a lot simpler, and I think that should be enough ?

Actually it's possible we will never hit the case: the event context is calculated by another resident server in a `make_event` remote call. It can contain local state (if a message was send to the resident server from the joining-already-joined server in the meantime), but unsure if/how this plays out with the unpartial state.

I think we should just retry the whole `make/send_join` in case of and that should be enough, in the same principle as 54c012c5a8722725cf104fa6205f253b5b9b0192.

**Comment by MatMaul** [comment]
> > However I am not sure this is really useful: retrying the remote join is a lot simpler, and I think that should be enough ?
> 
> Either approach (local or remote) sounds fine to me. Since this is already a race, it doesn't really matter which outcome you go for — so quite happy for you to choose the most simple and well-reasoned approach.

I tried the retry remote join instead, and it seems to end up with even more subtle changes because of the `else` and `finally` clause of the impacted `try`. I decide to keep it the way it is now (fallback to local join).

File: synapse/handlers/federation.py
Comment (squahtx) [reply synapse/handlers/federation.py:636]: There's still a race:
1. We enter `do_invite_join()`.
2. The room is un-partial stated in the background.
3. We read the partial state flag and see that it's not present.
4. We then proceed to re-partial state the room (!) which breaks a key assumption.

I propose we introduce a lock that guarantees that no other code may toggle the partial stated-ness of the room while the lock is held. That is, when joining:
1. We take the lock for the room
2. We do a partial send-join when i) we aren't in the room or ii) the room is partial stated. Corollary: we do a full join when we are in the room and it's not partial stated. This is the same condition as mentioned in another comment.
3. We persist the join, which should no longer be able to raise a `PartialStateConflictError`
4. We release the lock right after the code that kicks off the sync.

When syncing, we take the lock when we're about to finish and un-partial state the room.

We can use https://github.com/matrix-org/synapse/blob/develop/synapse/storage/databases/main/lock.py for the locking, but it needs an `acquire()` method to be implemented (keep attempting `try_acquire()` and sleeping? not ideal but implementing it properly with replication is a fair bit of work)

TODO: check in #synapse-devs if this makes sense.

File: synapse/handlers/federation.py
Comment (squahtx) [reply synapse/handlers/federation.py:669]: The proposal with a lock in another comment should eliminate `PartialStateConflictError`s and make this comment moot.

File: synapse/handlers/federation.py
Comment (squahtx) [reply synapse/handlers/federation.py:636]: @DMRobertson Heads up that some of the above may impact leaves while partial stated. I'm not entirely sure yet.

**Review by squahtx — COMMENTED** [review]
I did some thinking about the races and made some proposals for addressing them. I reckon introducing a lock around un-partial stating of rooms will fix them, though it's worth discussing it in #synapse-devs first. Sorry about the PR churn.

File: synapse/handlers/event_auth.py
Comment (DMRobertson) [inline synapse/handlers/event_auth.py:223]: I can see that we're only using the membership in this function; this looks correct. Any particular reason for making this change though?

If we're making these changes, could we change the docstring to explain what `prev_membership=None` means? I would guess this means that they've never been a member of the room in question?

File: synapse/handlers/federation.py
Comment (DMRobertson) [inline synapse/handlers/federation.py:568]: I can see this is taken from the comment below, but it hard to understand when this occurs. Is the point that

1. HS is partially joined to the room
2. This function gets called
3. We wait for a response from the remote HS
4. The resync for this room completes; the room is now no longer partially stated
5. We get a response from (3); but we should discard it and create a local join ourselves.

File: synapse/handlers/federation.py
Comment (DMRobertson) [reply synapse/handlers/federation.py:636]: > Maybe the partial state flag should imply that we are currently trying to sync partial state

FWIW this was my internal model for the meaning of that flag.

I'm not quite following all the moving changes here. I might have to pick your brains tomorrow

File: synapse/handlers/room_member.py
Comment (DMRobertson) [inline synapse/handlers/room_member.py:849]: Not fully sure what you're getting at here. Are you trying to enumerate the ways in which we use `state_before_join` and argue that it's okay that if this is only partial state?

File: synapse/handlers/room_member.py
Comment (DMRobertson) [inline synapse/handlers/room_member.py:1686]: I think this means that #12802 is going to have to ensure this returns False as soon as the last person has left? Feels like there's race potential

File: synapse/handlers/federation.py
Comment (squahtx) [reply synapse/handlers/federation.py:636]: > Maybe the partial state flag should imply that we are currently trying to sync partial state

We now think that the partial state flag just means that there are partial state events in the room and does not imply that there is a partial state sync running (because the partial state sync may have failed). It also does not imply that the local user currently has a member in the room.

**Comment by squahtx** [comment]
I tried to address the issues with restricted join in #14882, but it seems to have ended up second join-proofing `do_invite_join` and replacing that part of this PR. Please ignore all my comments about that method if #14882 lands.

File: synapse/handlers/federation.py
Comment (squahtx) [reply synapse/handlers/federation.py:575]: That's right (or was right prior to #14882, which guarantees no `PartialStateConflictError`s).

File: synapse/handlers/federation.py
Comment (squahtx) [reply synapse/handlers/federation.py]: Done as part of#14882.

File: synapse/handlers/federation.py
Comment (squahtx) [reply synapse/handlers/federation.py]: The proposed locking was implemented in #14882.

File: synapse/handlers/federation.py
Comment (squahtx) [reply synapse/handlers/federation.py]: #14882 implements the proposed lock, so we can no longer observe a `PartialStateConflictError`.

File: synapse/storage/databases/main/event_federation.py
Comment (squahtx) [reply synapse/storage/databases/main/event_federation.py]: #14882 implements that membership check proposed in the other review comment, which replaces this filter.

**Comment by squahtx** [comment]
Now that #14882 has landed, I've gone and resolved the merge conflicts.

All the PR comments plus changes for `do_invite_join` have gone into #14882, so that method is now compatible with second joins.
Also `do_invite_join` will no longer raise a PartialStateConflictError because there is locking now.

File: synapse/handlers/room_member.py
Comment (squahtx) [inline synapse/handlers/room_member.py:905]: I think knocking (line +1052) is going to be broken in that we can end up sending a knock event from a user with a ban we don't know about.
We can't exactly await full state, since we're holding the room membership lock. Any thoughts on what we should do in this case?

**Review by DMRobertson — COMMENTED** [review]
I think this LGTM and we should land it, provided that we have reasonable complement coverage to land with it.

File: synapse/handlers/room_member.py
Comment (DMRobertson) [reply synapse/handlers/room_member.py]: Many thanks, this is much appreciated.

File: synapse/handlers/room_member.py
Comment (DMRobertson) [reply synapse/handlers/room_member.py]: I think(?) you've addressed this by making it the caller's responsibility to interpret the return value of this function.

**Comment by DMRobertson** [comment]
Also: is the description of this PR still accurate to what it does, or should we update it?

File: synapse/handlers/room_member.py
Comment (squahtx) [reply synapse/handlers/room_member.py:905]: It's a bit different. Previously, we would do a remote /make_knock+/send_knock and the remote homeserver would tell us to go away we were banned.
Now, since we are in the room but do not have full state, we can add a local knock into the DAG that would otherwise be rejected and cite it as a prev/auth event.

File: synapse/handlers/federation_event.py
Comment (squahtx) [reply synapse/handlers/federation_event.py]: Updated the comment

File: synapse/handlers/federation_event.py
Comment (squahtx) [reply synapse/handlers/federation_event.py:559]: There doesn't appear to be such a check anywhere.

File: synapse/handlers/room_member.py
Comment (squahtx) [reply synapse/handlers/room_member.py:1153]: Added a test as part of https://github.com/matrix-org/complement/pull/567. Synapse develop currently blocks and fails the test.

**Comment by squahtx** [comment]
> Also: is the description of this PR still accurate to what it does, or should we update it?

Er, let's update it.

File: synapse/handlers/room_member.py
Comment (DMRobertson) [reply synapse/handlers/room_member.py:905]: Hmm. Would it be worth requiring knocks to go over federation during the resync? (Not necessarily in this PR)

Comment (squahtx) [reply synapse/handlers/room_member.py:905]: I'm not wholly convinced the remote knock code is suitable. It persists the knock event as an outlier. I'm not sure what happens if we then receive the knock event over federation.

Comment (DMRobertson) [reply synapse/handlers/room_member.py:905]: I'm inclined to land this as-is and kick any knocking cans down the road.

**Review by DMRobertson — APPROVED** [review]

**Pull request merged**
